### PR TITLE
Fix function calling for streaming

### LINKED ISSUE #168: Function calling not working as expected when streaming
**Describe the bug**
When using function calling along with streaming, function calls seem to be ignored as the model passes the function arguments in chunks. 

**To Reproduce**
1. Initialize an openai assistant with streaming set to `true`
2. Bind a function to the assistant
3. Pass a prompt to invoke the function

**Expected behavior**
Function call chunks should be collected and call once complete.

### PR DESCRIPTION
**Pull request by akshaylb**
# Pull Request Template

## Description

When using the streaming option with a tool binding, openAI returns function arguments in separate chunked responses. 
The current implementation assumed all function arguments are returned in one response and thus the function call kept failing due to it not satisfying the json schema requirements.

Fixes #168

- Added two session variables `allToolCalls` and `currentToolCall` to concatenate the incoming function response arguments, only terminating once a new function call with a new unique id comes in the request or the `EOS` is hit. 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- Defined a function and bound it to the LLM object. 
- Set tool choice to forcefully call the function. 
- Did the above steps with 2 function calls as well to simulate parallel function calls, but could not simulate it no matter how much I tried 😞 

## Checklist:

- [x] I have read the [Contributing documentation](https://github.com/henomis/lingoose/blob/main/CONTRIBUTING.md).
- [x] I have read the [Code of conduct documentation](https://github.com/henomis/lingoose/blob/main/CODE_OF_CONDUCT.md).
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by henomis** [comment]
@akshaylb can you please add an example as `examples/llm/openai/thread-stream` copying the `thread` and adding `WithStream()` to the llm?

**Comment by henomis** [comment]
@akshaylb any news on this?

**Comment by akshaylb** [comment]
Apologies for the delay @henomis 
Updated the example file.

**Review by henomis — CHANGES_REQUESTED** [review]

File: llm/openai/openai.go [hunk 1]
```diff
@@ -239,7 +245,16 @@ func (o *OpenAI) stream(
 		}
 
 		if response.Choices[0].FinishReason == "tool_calls" || len(response.Choices[0].Delta.ToolCalls) > 0 {
```
[the remark this answers is not in the document]
Comment (akshaylb) [reply llm/openai/openai.go]: Updated!

File: llm/openai/openai.go [hunk 2]
```diff
@@ -239,7 +245,16 @@ func (o *OpenAI) stream(
...
+			if len(response.Choices[0].Delta.ToolCalls) > 0 {
+				if response.Choices[0].Delta.ToolCalls[0].ID != "" {
+					if currentToolCall.ID != "" {
+						allToolCalls = append(allToolCalls, currentToolCall)
+					}
+					currentToolCall = response.Choices[0].Delta.ToolCalls[0]
+				} else {
+					currentToolCall.Function.Arguments += response.Choices[0].Delta.ToolCalls[0].Function.Arguments
+				}
+			}
```
[the remark this answers is not in the document]
Comment (akshaylb) [reply llm/openai/openai.go]: Moved logic to function!

**Review by henomis — CHANGES_REQUESTED** [review]

File: llm/openai/openai.go [hunk 3]
```diff
@@ -206,6 +206,22 @@ func (o *OpenAI) Generate(ctx context.Context, t *thread.Thread) error {
 	return nil
 }
 
+func isStreamToolCallResponse(response *openai.ChatCompletionStreamResponse) bool {
+	return response.Choices[0].FinishReason == openai.FinishReasonToolCalls || len(response.Choices[0].Delta.ToolCalls) > 0
+}
+
+func consumeToolCallResponse(response *openai.ChatCompletionStreamResponse, currentToolCall *openai.ToolCall) (updatedToolCall openai.ToolCall, isNewTool bool) {
```
[the remark this answers is not in the document]
Comment (akshaylb) [reply llm/openai/openai.go]: Hi @henomis 
Updated this change as well.

File: llm/openai/openai.go [hunk 4]
```diff
@@ -221,25 +239,37 @@ func (o *OpenAI) stream(
 
 	var messages []*thread.Message
 	var content string
+	var allToolCalls []openai.ToolCall
```
Comment (henomis) [inline llm/openai/openai.go:242]: To reduce complexity we have to put 242-278 to a separated method.

**Review by henomis — CHANGES_REQUESTED** [review]

File: llm/openai/openai.go (hunk 4 above)
Comment (akshaylb) [reply llm/openai/openai.go:258]: @henomis , I tried to move the entire block and tried running `golangci-lint` but it still threw the same error. 
I have thus moved only the error handling logic to a separate function to reduce complexity. 
`golangci-lint` doesnt throw any errors now.

Comment (henomis) [reply llm/openai/openai.go:258]: Cool! Let me reserve a slot of time for the last review, and we should be ready to merge it. Thank you so much for your effort! :rocket:

File: llm/openai/openai.go [hunk 5]
```diff
@@ -206,6 +206,40 @@ func (o *OpenAI) Generate(ctx context.Context, t *thread.Thread) error {
 	return nil
 }
 
+func isStreamToolCallResponse(response *openai.ChatCompletionStreamResponse) bool {
+	return response.Choices[0].FinishReason == openai.FinishReasonToolCalls || len(response.Choices[0].Delta.ToolCalls) > 0
+}
+
+func consumeToolCallResponse(response *openai.ChatCompletionStreamResponse,
+	currentToolCall *openai.ToolCall) (updatedToolCall openai.ToolCall, isNewTool bool) {
```
Comment (henomis) [inline llm/openai/openai.go:214]: Let's call it `handleStreamToolCallResponse` and put parameters on new lines.

```suggestion
func handleStreamToolCallResponse(
	response *openai.ChatCompletionStreamResponse,
	currentToolCall *openai.ToolCall,
) (updatedToolCall openai.ToolCall, isNewTool bool) {
```

File: llm/openai/openai.go [hunk 6]
```diff
@@ -206,6 +206,40 @@ func (o *OpenAI) Generate(ctx context.Context, t *thread.Thread) error {
...
+		isNewTool = true
+		updatedToolCall = response.Choices[0].Delta.ToolCalls[0]
+	} else {
+		currentToolCall.Function.Arguments += response.Choices[0].Delta.ToolCalls[0].Function.Arguments
+	}
+	return
+}
+
+func (o *OpenAI) processErrorResponse(messages []*thread.Message, content string,
+	currentToolCall *openai.ToolCall, allToolCalls []openai.ToolCall) []*thread.Message {
```
Comment (henomis) [inline llm/openai/openai.go:228]: Let's call it `handleEndOfStream` 'cause even though `EOF` is an error it means that there is no more data on the stream.

```suggestion
func (o *OpenAI) handleEndOfStream(
	messages []*thread.Message,
	content string,
	currentToolCall *openai.ToolCall,
	allToolCalls []openai.ToolCall,
) []*thread.Message {
```

File: llm/openai/openai.go [hunk 7]
```diff
@@ -219,27 +253,29 @@ func (o *OpenAI) stream(
...
 		response, errRecv := stream.Recv()
 		if errors.Is(errRecv, io.EOF) {
-			o.streamCallbackFn(EOS)
-
-			if len(content) > 0 {
-				messages = append(messages, thread.NewAssistantMessage().AddContent(
-					thread.NewTextContent(content),
-				))
-			}
+			messages = o.processErrorResponse(messages, content, &currentToolCall, allToolCalls)
```
Comment (henomis) [inline llm/openai/openai.go:263]: use the new function name

```suggestion
			messages = o.handleEndOfStream(messages, content, &currentToolCall, allToolCalls)
```

File: llm/openai/openai.go [hunk 8]
```diff
@@ -219,27 +253,29 @@ func (o *OpenAI) stream(
...
 		}
 
 		if len(response.Choices) == 0 {
 			return fmt.Errorf("%w: no choices returned", ErrOpenAIChat)
 		}
 
-		if response.Choices[0].FinishReason == "tool_calls" || len(response.Choices[0].Delta.ToolCalls) > 0 {
-			messages = append(messages, o.callTools(response.Choices[0].Delta.ToolCalls)...)
+		if isStreamToolCallResponse(&response) {
+			updatedToolCall, isNewTool := consumeToolCallResponse(&response, &currentToolCall)
```
Comment (henomis) [inline llm/openai/openai.go:272]: Use the new function name.

```suggestion
			updatedToolCall, isNewTool := handleStreamToolCallResponse(&response, &currentToolCall)
```

**Review by henomis — CHANGES_REQUESTED** [review]

**Comment by akshaylb** [comment]
@henomis changes completed.

**Review by henomis — APPROVED** [review]

**Pull request merged**